File: chrome/android/features/autofill_assistant/java/src/org/chromium/chrome/browser/autofill_assistant/header/AssistantHeaderModel.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.autofill_assistant.header;

import android.support.annotation.VisibleForTesting;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.chrome.browser.autofill_assistant.carousel.AssistantChip;
import org.chromium.chrome.browser.autofill_assistant.generic_ui.AssistantDrawable;
import org.chromium.ui.modelutil.PropertyModel;

import java.util.ArrayList;
import java.util.List;

/**
 * State for the header of the Autofill Assistant.
 */
@JNINamespace("autofill_assistant")
public class AssistantHeaderModel extends PropertyModel {
    public static final WritableObjectPropertyKey<List<AssistantChip>> CHIPS =
            new WritableObjectPropertyKey<>();

    public static final WritableObjectPropertyKey<String> STATUS_MESSAGE =
            new WritableObjectPropertyKey<>();

    public static final WritableObjectPropertyKey<String> BUBBLE_MESSAGE =
            new WritableObjectPropertyKey<>();

    public static final WritableIntPropertyKey PROGRESS = new WritableIntPropertyKey();

    public static final WritableIntPropertyKey PROGRESS_ACTIVE_STEP = new WritableIntPropertyKey();

    public static final WritableBooleanPropertyKey PROGRESS_BAR_ERROR =
            new WritableBooleanPropertyKey();

    public static final WritableBooleanPropertyKey PROGRESS_VISIBLE =
            new WritableBooleanPropertyKey();

    public static final WritableBooleanPropertyKey USE_STEP_PROGRESS_BAR =
            new WritableBooleanPropertyKey();

    public static final WritableObjectPropertyKey<List<AssistantDrawable>> STEP_PROGRESS_BAR_ICONS =
            new WritableObjectPropertyKey<>();

    static final WritableBooleanPropertyKey SPIN_POODLE = new WritableBooleanPropertyKey();

    public static final WritableObjectPropertyKey<Runnable> FEEDBACK_BUTTON_CALLBACK =
            new WritableObjectPropertyKey<>();

    public static final WritableBooleanPropertyKey CHIPS_VISIBLE = new WritableBooleanPropertyKey();

    public static final WritableBooleanPropertyKey DISABLE_ANIMATIONS_FOR_TESTING =
            new WritableBooleanPropertyKey();

    public AssistantHeaderModel() {
        super(STATUS_MESSAGE, BUBBLE_MESSAGE, PROGRESS, PROGRESS_ACTIVE_STEP, PROGRESS_BAR_ERROR,
                PROGRESS_VISIBLE, USE_STEP_PROGRESS_BAR, STEP_PROGRESS_BAR_ICONS, SPIN_POODLE,
                FEEDBACK_BUTTON_CALLBACK, CHIPS, CHIPS_VISIBLE, DISABLE_ANIMATIONS_FOR_TESTING);
        set(CHIPS, new ArrayList<>());
        set(PROGRESS_VISIBLE, true);
    }

    @CalledByNative
    private void setStatusMessage(String statusMessage) {
        set(STATUS_MESSAGE, statusMessage);
    }

    @CalledByNative
    private void setBubbleMessage(String bubbleMessage) {
        set(BUBBLE_MESSAGE, bubbleMessage);
    }

    @CalledByNative
    private void setProgress(int progress) {
        set(PROGRESS, progress);
    }

    @CalledByNative
    private void setProgressActiveStep(int activeStep) {
        set(PROGRESS_ACTIVE_STEP, activeStep);
    }

    @CalledByNative
    private void setProgressBarErrorState(boolean error) {
        set(PROGRESS_BAR_ERROR, error);
    }

    @CalledByNative
    private void setProgressVisible(boolean visible) {
        set(PROGRESS_VISIBLE, visible);
    }

    @CalledByNative
    private void setUseStepProgressBar(boolean useNewProgressBar) {
        set(USE_STEP_PROGRESS_BAR, useNewProgressBar);
    }

    @CalledByNative
    private static List<AssistantDrawable> createIconList() {
        return new ArrayList<>();
    }

    @CalledByNative
    private static void addStepProgressBarIcon(
            List<AssistantDrawable> icons, AssistantDrawable icon) {
        icons.add(icon);
    }

    @CalledByNative
    private void setStepProgressBarIcons(List<AssistantDrawable> icons) {
        set(STEP_PROGRESS_BAR_ICONS, icons);
        // Reset progress bar entries.
        set(PROGRESS_ACTIVE_STEP, -1);
        set(PROGRESS_BAR_ERROR, false);
    }

    @CalledByNative
    private void setSpinPoodle(boolean enabled) {
        set(SPIN_POODLE, enabled);
    }

    @CalledByNative
    private void setDelegate(AssistantHeaderDelegate delegate) {
        set(FEEDBACK_BUTTON_CALLBACK, delegate::onFeedbackButtonClicked);
    }

    @CalledByNative
    private void setDisableAnimations(boolean disableAnimations) {
        set(DISABLE_ANIMATIONS_FOR_TESTING, disableAnimations);
    }

    @CalledByNative
    @VisibleForTesting
    public void setChips(List<AssistantChip> chips) {
        // Move last chip (cancel) to first position. For legacy reasons, native builds this list
        // such that the cancel chip is last, but the regular carousel will show it in the left-most
        // position and the header should mirror this.
        if (chips.size() > 1) {
            chips.add(0, chips.remove(chips.size() - 1));
        }
        set(CHIPS, chips);
    }
}

File: chrome/browser/android/autofill_assistant/assistant_header_model.cc
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/autofill_assistant/assistant_header_model.h"

#include "base/android/jni_string.h"
#include "chrome/android/features/autofill_assistant/jni_headers/AssistantHeaderModel_jni.h"
#include "chrome/browser/android/autofill_assistant/ui_controller_android.h"
#include "chrome/browser/android/autofill_assistant/ui_controller_android_utils.h"

using base::android::AttachCurrentThread;

namespace autofill_assistant {

AssistantHeaderModel::AssistantHeaderModel(
    const base::android::ScopedJavaLocalRef<jobject>& jmodel)
    : jmodel_(jmodel) {}

AssistantHeaderModel::~AssistantHeaderModel() = default;

void AssistantHeaderModel::SetDelegate(
    const AssistantHeaderDelegate& delegate) {
  Java_AssistantHeaderModel_setDelegate(AttachCurrentThread(), jmodel_,
                                        delegate.GetJavaObject());
}

void AssistantHeaderModel::SetStatusMessage(const std::string& status_message) {
  JNIEnv* env = AttachCurrentThread();
  Java_AssistantHeaderModel_setStatusMessage(
      env, jmodel_,
      base::android::ConvertUTF8ToJavaString(env, status_message));
}

void AssistantHeaderModel::SetBubbleMessage(const std::string& bubble_message) {
  JNIEnv* env = AttachCurrentThread();
  Java_AssistantHeaderModel_setBubbleMessage(
      env, jmodel_,
      base::android::ConvertUTF8ToJavaString(env, bubble_message));
}

void AssistantHeaderModel::SetProgress(int progress) {
  Java_AssistantHeaderModel_setProgress(AttachCurrentThread(), jmodel_,
                                        progress);
}

void AssistantHeaderModel::SetProgressActiveStep(int active_step) {
  Java_AssistantHeaderModel_setProgressActiveStep(AttachCurrentThread(),
                                                  jmodel_, active_step);
}

void AssistantHeaderModel::SetProgressVisible(bool visible) {
  Java_AssistantHeaderModel_setProgressVisible(AttachCurrentThread(), jmodel_,
                                               visible);
}

void AssistantHeaderModel::SetProgressBarErrorState(bool error) {
  Java_AssistantHeaderModel_setProgressBarErrorState(AttachCurrentThread(),
                                                     jmodel_, error);
}

void AssistantHeaderModel::SetStepProgressBarConfiguration(
    const ShowProgressBarProto::StepProgressBarConfiguration& configuration,
    const base::android::ScopedJavaLocalRef<jobject>& jcontext) {
  JNIEnv* env = AttachCurrentThread();
  Java_AssistantHeaderModel_setUseStepProgressBar(
      env, jmodel_, configuration.use_step_progress_bar());
  if (!configuration.annotated_step_icons().empty()) {
    auto jlist = Java_AssistantHeaderModel_createIconList(env);
    for (const auto& icon : configuration.annotated_step_icons()) {
      Java_AssistantHeaderModel_addStepProgressBarIcon(
          env, jlist,
          ui_controller_android_utils::CreateJavaDrawable(env, jcontext,
                                                          icon.icon()));
    }
    Java_AssistantHeaderModel_setStepProgressBarIcons(env, jmodel_, jlist);
  }
}

void AssistantHeaderModel::SetSpinPoodle(bool enabled) {
  Java_AssistantHeaderModel_setSpinPoodle(AttachCurrentThread(), jmodel_,
                                          enabled);
}

void AssistantHeaderModel::SetChips(
    const base::android::ScopedJavaLocalRef<jobject>& jchips) {
  Java_AssistantHeaderModel_setChips(AttachCurrentThread(), jmodel_, jchips);
}

void AssistantHeaderModel::SetDisableAnimations(bool disable_animations) {
  Java_AssistantHeaderModel_setDisableAnimations(AttachCurrentThread(), jmodel_,
                                                 disable_animations);
}

base::android::ScopedJavaGlobalRef<jobject>
AssistantHeaderModel::GetJavaObject() const {
  return jmodel_;
}

}  // namespace autofill_assistant
